lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -322,6 +322,104 @@ void SPIRVRegularizeLLVMBase::expandSYCLTypeUsing(Module *M) {
     expandVIDWithSYCLTypeByValComp(F);
 }
 
+// In this function, we handle two conversion operations
+// 1. fptoui.sat.iX.fY (X is not 8,16,32,64; Y is 32 or 64)
+// 2. fptosi.sat.iX.fY (X is not 8,16,32,64; Y is 32 or 64)
+// Such non-standard integer types cannot be handled in SPIR-V. Hence, they
+// will be promoted to
+// 1. fptoui.sat.i64.fY (Y is 32 or 64)
+// 2. fptosi.sat.i64.fY (Y is 32 or 64)
+// However, LLVM documentation requires the following rules to be obeyed.
+// Rule 1: If the argument is any NaN, zero is returned.
+// Rule 2: If the argument is smaller than the smallest representable
+// (un)signed integer of the result type, the smallest representable
+// (un)signed integer is returned.
+// Rule 3: If the argument is larger than the largest representable (un)signed
+// integer of the result type, the largest representable (un)signed integer is
+// returned.
+// Rule 4: Otherwise, the result of rounding the argument towards zero is
+// returned.
+// Rules 1 & 4 are preserved when promoting iX to i64. For preserving Rule 2
+// and Rule 3, we saturate the result of the promoted instruction based on
+// original integer type (iX)
+// Example:
+// Input:
+// %0 = call i2 @llvm.fptosi.sat.i2.f32(float %input)
+// %1 = sext i32 %0
```

Comment:
Would optimizer always extend generated integer and never use the original type?

Reply:
Yes.